File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/FairnessLayFunction.java
package cz.fb.manaus.reactor.profit.progress.function;

import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.reactor.price.FairnessPolynomialCalculator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.OptionalDouble;

@Component
public class FairnessLayFunction implements ProgressFunction {

    @Autowired
    private FairnessPolynomialCalculator calculator;

    @Override
    public OptionalDouble apply(SettledBet bet) {
        var marketPrices = bet.getBetAction().getMarketPrices();
        return calculator.getFairness(marketPrices.getWinnerCount(), marketPrices.getBestPrices(Side.LAY));
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/progress/function/FairnessLayFunction.kt
package cz.fb.manaus.reactor.profit.progress.function

import cz.fb.manaus.core.model.RealizedBet
import cz.fb.manaus.core.model.Side
import cz.fb.manaus.reactor.price.FairnessPolynomialCalculator
import cz.fb.manaus.reactor.price.getBestPrices
import org.springframework.stereotype.Component

@Component
object FairnessLayFunction : ProgressFunction {

    override fun invoke(bet: RealizedBet): Double? {
        val marketPrices = bet.betAction.runnerPrices
        return FairnessPolynomialCalculator.getFairness(1, getBestPrices(marketPrices, Side.LAY))
    }

}
